src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -897,25 +953,27 @@ private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request
                     (remaining.Length == 0 ? FrameFlags.EndHeaders : FrameFlags.None) |
                     (request.Content == null ? FrameFlags.EndStream : FrameFlags.None);
 
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                await StartWriteAsync(FrameHeader.Size + current.Length, default, false).ConfigureAwait(false);
+
                 WriteFrameHeader(new FrameHeader(current.Length, FrameType.Headers, flags, streamId));
                 current.CopyTo(_outgoingBuffer.AvailableMemory);
                 _outgoingBuffer.Commit(current.Length);
 
-                await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                await FinishWriteAsync(remaining.Length == 0, cancellationToken, false).ConfigureAwait(false);
 
                 while (remaining.Length > 0)
                 {
                     (current, remaining) = SplitBuffer(remaining, FrameHeader.MaxLength);
 
                     flags = (remaining.Length == 0 ? FrameFlags.EndHeaders : FrameFlags.None);
 
-                    _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size + current.Length);
+                    await StartWriteAsync(FrameHeader.Size + current.Length, default, false).ConfigureAwait(false);
+
                     WriteFrameHeader(new FrameHeader(current.Length, FrameType.Continuation, flags, streamId));
                     current.CopyTo(_outgoingBuffer.AvailableMemory);
                     _outgoingBuffer.Commit(current.Length);
 
-                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                    await FinishWriteAsync(remaining.Length == 0, cancellationToken, false).ConfigureAwait(false);
```

Comment:
I think you only need to force the flush in the EndStream case. If it's not EndStream then there will be body frames to follow.

Reply:
This will be fixed in the next iteration. Thanks.